packages/fluentui/react-northstar/src/components/Tree/hooks/useTreeActiveState.tsx

```diff
@@ -0,0 +1,105 @@
+import * as React from 'react';
+import { useAutoControlled } from '@fluentui/react-bindings';
+import { TreeProps } from '../Tree';
+import { useStableProps } from './useStableProps';
+import { FlatTree } from './flattenTree';
+import { findIndex, removeItemAtIndex } from './utils';
+import * as _ from 'lodash';
+
+export interface TreeActiveState {
+  activeItemIds: string[];
+  flatTree: FlatTree;
+  toggleActive: (ids: string[], e: React.SyntheticEvent) => void;
+}
+
+export function useTreeActiveState(
+  props: Pick<TreeProps, 'defaultActiveItemIds' | 'activeItemIds' | 'exclusive'>,
+  flatTree: FlatTree,
+): TreeActiveState {
+  const initialActiveItemIds = React.useMemo(() => {
+    const initalValue = [];
+    Object.keys(flatTree).forEach(key => {
+      if (flatTree[key].expanded) {
+        initalValue.push(key);
+      }
+    });
+    return initalValue;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []); // initialValue only needs to be computed on mount
+
+  const [activeItemIds, setActiveItemIdsState] = useAutoControlled<string[]>({
+    defaultValue: props.defaultActiveItemIds,
+    value: props.activeItemIds,
+    initialValue: initialActiveItemIds,
+  });
+
+  // compute a new flattened tree based on activeItemIds state
+  const updatedFlatTree = React.useMemo(() => {
+    const updatedFlatTree = { ...flatTree };
+    Object.keys(updatedFlatTree).forEach(key => {
+      if (updatedFlatTree[key].expanded) {
+        updatedFlatTree[key].expanded = false;
+      }
+    });
+    activeItemIds.forEach(activeId => {
+      updatedFlatTree[activeId].expanded = true;
+    });
+    return updatedFlatTree;
+  }, [activeItemIds, flatTree]);
+
+  const toggleActiveOnOneId = React.useCallback(
+    (ids: string[], id: string): string[] => {
+      if (!updatedFlatTree[id]?.hasSubtree) {
+        // leaf node does not have the concept of active/inactive
+        return ids;
+      }
+
+      let result: string[];
+      const index = findIndex(ids, id);
+      if (index >= 0) {
+        result = removeItemAtIndex(ids, index);
+      } else {
+        if (props.exclusive) {
+          // need to collapse everything else, except id and its ancestors
+          const ancestors = getAncestorsIds(updatedFlatTree, id);
+          return [...ancestors, id];
+        }
+        return [...ids, id];
+      }
+
+      return result;
+    },
+    [props.exclusive, updatedFlatTree],
```

Comment:
**(!)** Having a dependency there on `updatedFlatTree` is a bad idea as `toggleActive` is a part of `contextValue`

Reply:
I understand. However I'm not sure what would be a good way to avoid so, as we are changing the flatTree based on the active ids change. Let's talk about this tomorrow

Reply:
Resolved  by `useGetItemById`